db/model/Variable.ts

```diff
@@ -366,24 +434,75 @@ export const dataAsDF = async (
     )
 
     if (df.height == 0) {
-        df = pl.DataFrame({
-            variableId: [],
-            entityId: [],
-            entityName: [],
-            entityCode: [],
-            year: [],
-            value: [],
-        })
+        return emptyDataDF()
     }
 
-    return df.select(
-        pl.col("variableId").cast(pl.Int32),
-        pl.col("entityId").cast(pl.Int32),
-        pl.col("entityName").cast(pl.Utf8),
-        pl.col("entityCode").cast(pl.Utf8),
-        pl.col("year").cast(pl.Int32),
-        pl.col("value").cast(pl.Utf8)
+    return _castDataDF(df)
+}
+
+export const dataAsDF = async (
+    variableIds: OwidVariableId[]
+): Promise<pl.DataFrame> => {
+    // return variables values as a DataFrame from either S3 or data_values table
+    const rows = await db.queryMysql(
+        `
+    SELECT
+        id,
+        dataPath
+    FROM variables
+    WHERE id in (?)
+    `,
+        [variableIds]
     )
+
+    // variables with dataPath are stored in S3 and variables without it
+    // are stored in data_values table
+    const variableIdsWithDataPath = rows
+        .filter((row: any) => row.dataPath)
+        .map((row: any) => row.id)
+
+    const variableIdsWithoutDataPath = rows
+        .filter((row: any) => !row.dataPath)
+        .map((row: any) => row.id)
+
+    return pl.concat([
+        await _dataAsDFfromMySQL(variableIdsWithoutDataPath),
+        await _dataAsDFfromS3(variableIdsWithDataPath),
+    ])
+}
+
+export const getOwidVariableDataPath = async (
+    variableId: OwidVariableId
+): Promise<string | undefined> => {
+    const row = await db.mysqlFirst(
+        `SELECT dataPath FROM variables WHERE id = ?`,
+        [variableId]
+    )
+    return row.dataPath
+}
+
+// limit number of concurrent requests to 50 when fetching values from S3
+const httpsAgent = new https.Agent({
+    keepAlive: true,
+    maxSockets: 50,
+})
+
+export const fetchS3Values = async (
+    variableId: OwidVariableId
+): Promise<OwidVariableMixedData> => {
+    const dataPath = await getOwidVariableDataPath(variableId)
+    if (!dataPath) {
+        throw new Error(`Missing dataPath for variable ${variableId}`)
+    }
+    return fetchS3ValuesByPath(dataPath)
+}
+
+export const fetchS3ValuesByPath = async (
+    dataPath: string
+): Promise<OwidVariableMixedData> => {
+    return (await (
```

Comment:
You don't need the outer await here (but then you need to change the type to a promise like in the return type). In this case I think that is a bit nicer to read